File: I1_MVPAndRetrofit/app/src/main/java/com/example/tsvetelinastoyanova/weatherreportrevisited/threadhandlers/AppExecutors.java
package com.example.tsvetelinastoyanova.weatherreportrevisited.threadhandlers;

import android.os.Handler;
import android.os.Looper;
import android.support.annotation.NonNull;
import android.support.annotation.VisibleForTesting;

import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

/**
 * Global executor pools for the whole application.
 * <p>
 * Grouping tasks like this avoids the effects of task starvation (e.g. disk reads don't wait behind
 * webservice requests).
 */
public class AppExecutors {

    private static final int THREAD_COUNT = 1;

    private final Executor databaseIO;

    private final Executor networkIO;

    private final Executor mainThread;

    @VisibleForTesting
    AppExecutors(Executor databaseIO, Executor networkIO, Executor mainThread) {
        this.databaseIO = databaseIO;
        this.networkIO = networkIO;
        this.mainThread = mainThread;
    }

    public AppExecutors() {
        this(new DatabaseIOThreadExecutor(), Executors.newFixedThreadPool(THREAD_COUNT),
                new MainThreadExecutor());
    }

    public Executor databaseIO() {
        return databaseIO;
    }

    public Executor networkIO() {
        return networkIO;
    }

    public Executor mainThread() {
        return mainThread;
    }

    private static class MainThreadExecutor implements Executor {
        private Handler mainThreadHandler = new Handler(Looper.getMainLooper());

        @Override
        public void execute(@NonNull Runnable command) {
            mainThreadHandler.post(command);
        }
    }
}

File: I1_MVPAndRetrofit_Kotlin/app/src/main/java/com/example/tsvetelinastoyanova/weatherapp/util/AppExecutors.kt
package com.example.tsvetelinastoyanova.weatherapp.util

import android.os.Handler
import android.os.Looper
import android.support.annotation.VisibleForTesting
import java.util.concurrent.Executor
import java.util.concurrent.Executors
/*

class AppExecutors @VisibleForTesting
internal constructor(private val databaseIO: Executor, private val networkIO: Executor, private val mainThread: Executor) {

    constructor() : this(DatabaseIOThreadExecutor(), Executors.newFixedThreadPool(THREAD_COUNT),
            MainThreadExecutor())

    fun databaseIO(): Executor {
        return databaseIO
    }

    fun networkIO(): Executor {
        return networkIO
    }

    fun mainThread(): Executor {
        return mainThread
    }

    private class MainThreadExecutor : Executor {
        private val mainThreadHandler = Handler(Looper.getMainLooper())

        override fun execute(command: Runnable) {
            mainThreadHandler.post(command)
        }
    }

    companion object {

        private val THREAD_COUNT = 3
    }
}*/